plugins/inputs/systemd_units/systemd_units.go

```diff
@@ -128,109 +119,80 @@ var subMap = map[string]int{
-	defaultUnitType = "service"
-	defaultPattern  = ""
+	defaultSubCommand = "list-units"
+	defaultTimeout    = config.Duration(time.Second)
+	defaultUnitType   = "service"
+	defaultPattern    = ""
 )
 
 func (*SystemdUnits) SampleConfig() string {
 	return sampleConfig
 }
 
-// Gather parses systemctl outputs and adds counters to the Accumulator
-func (s *SystemdUnits) Gather(acc telegraf.Accumulator) error {
-	out, err := s.systemctl(s.Timeout, s.UnitType, s.Pattern)
-	if err != nil {
-		return err
-	}
+func (s *SystemdUnits) Init() error {
+	var subCommandInfo *subCommandInfo
 
-	scanner := bufio.NewScanner(out)
-	for scanner.Scan() {
-		line := scanner.Text()
-
-		data := strings.Fields(line)
-		if len(data) < 4 {
-			acc.AddError(fmt.Errorf("Error parsing line (expected at least 4 fields): %s", line))
-			continue
-		}
-		name := data[0]
-		load := data[1]
-		active := data[2]
-		sub := data[3]
-		tags := map[string]string{
-			"name":   name,
-			"load":   load,
-			"active": active,
-			"sub":    sub,
-		}
-
-		var (
-			loadCode   int
-			activeCode int
-			subCode    int
-			ok         bool
-		)
-		if loadCode, ok = loadMap[load]; !ok {
-			acc.AddError(fmt.Errorf("Error parsing field 'load', value not in map: %s", load))
-			continue
-		}
-		if activeCode, ok = activeMap[active]; !ok {
-			acc.AddError(fmt.Errorf("Error parsing field 'active', value not in map: %s", active))
-			continue
-		}
-		if subCode, ok = subMap[sub]; !ok {
-			acc.AddError(fmt.Errorf("Error parsing field 'sub', value not in map: %s", sub))
-			continue
-		}
-		fields := map[string]interface{}{
-			"load_code":   loadCode,
-			"active_code": activeCode,
-			"sub_code":    subCode,
-		}
-
-		acc.AddFields(measurement, fields, tags)
+	switch s.SubCommand {
+	case "show":
+		subCommandInfo = initSubcommandShow()
+	case "list-units":
+		subCommandInfo = initSubcommandListUnits()
+	default:
+		return fmt.Errorf("invalid value for 'subcommand'")
 	}
 
+	// Save the parsing function for later and pre-compute the
+	// command line because it will not change.
+	s.resultParser = subCommandInfo.parseResult
+	s.commandParams = subCommandInfo.getParameters(s)
+
 	return nil
 }
 
-func setSystemctl(timeout config.Duration, unitType string, pattern string) (*bytes.Buffer, error) {
+func (s *SystemdUnits) Gather(acc telegraf.Accumulator) error {
 	// is systemctl available ?
 	systemctlPath, err := exec.LookPath("systemctl")
 	if err != nil {
-		return nil, err
-	}
-	// build parameters for systemctl call
-	params := []string{"list-units"}
-	// create patterns parameters if provided in config
-	if pattern != "" {
-		psplit := strings.SplitN(pattern, " ", -1)
-		params = append(params, psplit...)
+		return err
 	}
-	params = append(params,
-		"--all", "--plain",
-		// add type as configured in config
-		"--type="+unitType,
-		"--no-legend",
-	)
-	cmd := exec.Command(systemctlPath, params...)
+
+	cmd := exec.Command(systemctlPath, *s.commandParams...)
```

Comment:
How about building the command at `Init` time?

Reply:
The [go documentation](https://pkg.go.dev/os/exec#Cmd) states: "A Cmd cannot be reused after calling its Run, Output or CombinedOutput methods."
I thought it would not be possible to reuse the `Cmd` instance returned by `exec.Command`. Did I miss something?

Reply:
You are absolutely right. Sorry for the noise!